### PR TITLE
fix: restore process fails due to DROP constraints

### PR DESCRIPTION
**Pull request by spena**
### Description 
With [feat: add constraint to deny drop sources referenced by other CREATE_AS sourcesCLOSED](https://github.com/confluentinc/ksql/pull/6545) in 0.15 we introduced constraints that deny DROP statements when the object to drop is referenced by other objects. For example, in the following situation:

```
CREATE STREAM streamA { ... }  
...
CREATE STREAM streamB AS SELECT ... FROM streamA EMIT CHANGES; 
```

`streamA` cannot be dropped since `streamB` still references `streamA`. Before 0.15 the constraints had not existed and `streamA` could have been dropped.

This means, command topics that contain DROP statements from before 0.15 might block an upgrade to later versions since the constraint on the DROP statement would deny the DROP and throw the following exception:
```
io.confluent.ksql.util.KsqlStatementException: Cannot drop streamA.
The following streams and/or tables read from this source: [streamB].
You need to drop them before dropping streamA.
Statement: DROP STREAM streamA;
      at io.confluent.ksql.engine.EngineExecutor.executeDdl(EngineExecutor.java:610)
      at ...
```

This PR fixes the issue by ignoring the DROP constraints when the metastore is restoring. The constraints are still checked if the DROP is executed manually from the CLI or REST API.

Also, if the dropped source is created again during the metastore restore process, then the metastore adds the drop constraint again to be consistent to the drop constraint feature. However, if the CREATE is executed manually on the CLI or REST, the constraint is not added back.

### Testing done 
Added unit tests
Verified manually

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-engine/src/main/java/io/confluent/ksql/ddl/commands/DdlCommandExec.java [hunk 1]
```diff
@@ -52,31 +52,44 @@ public DdlCommandExec(final MutableMetaStore metaStore) {
...
         final String sql,
         final boolean withQuery,
-        final Set<SourceName> withQuerySources
+        final Set<SourceName> withQuerySources,
+        final boolean restoreInProgress
     ) {
       this.sql = Objects.requireNonNull(sql, "sql");
       this.withQuery = withQuery;
       this.withQuerySources = Objects.requireNonNull(withQuerySources, "withQuerySources");
+      this.restoreInProgress = Objects.requireNonNull(restoreInProgress, "restoreInProgress");
```
Comment (jnh5y) [inline ksqldb-engine/src/main/java/io/confluent/ksql/ddl/commands/DdlCommandExec.java:92]: nit: As a primitive, boolean cannot be null.  `withQuery` isn't checked.

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/RestoreCommandsCompactorTest.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+package io.confluent.ksql.rest.server.computation;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class RestoreCommandsCompactorTest {
```
Comment (jnh5y) [inline ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/RestoreCommandsCompactorTest.java:23]: Should this test be completed or removed?

File: ksqldb-metastore/src/test/java/io/confluent/ksql/metastore/MetaStoreImplTest.java [hunk 3]
```diff
@@ -263,6 +265,53 @@ public void shouldDeleteSourceAfterLinkedSourceIsDeleted() {
...
+    // Given:
+    metaStore.putSource(dataSource, false);
+    metaStore.putSource(dataSource1, false);
+    metaStore.addSourceReferences(dataSource1.getName(),
+        Collections.singleton(dataSource.getName()));
+
+    // force source delete if constraint exists
+    metaStore.deleteSource(dataSource.getName(), true);
+
+    // When:
```
Comment (jnh5y) [inline ksqldb-metastore/src/test/java/io/confluent/ksql/metastore/MetaStoreImplTest.java:301]: I'd be tempted to add checks before this to show that `dataSource` has been removed and that the constraint is not present.

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/RecoveryTest.java [hunk 4]
```diff
@@ -767,7 +859,8 @@ public void shouldRecoverDropWithSourceConstraintsFromOldMetastore() {
     assertThat(server1.ksqlEngine.getMetaStore().getAllDataSources(), hasKey(SourceName.of("B")));
 
     // Recovered server has only stream 'B'
```
Comment (jnh5y) [inline ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/RecoveryTest.java:861]: Update comment?

**Comment by jnh5y** [comment]
@spena looks reasonable.  Lemme know when you've had a chance to answer the questions and I'm happy to approve!

File: ksqldb-engine/src/main/java/io/confluent/ksql/ddl/commands/DdlCommandExec.java (hunk 1 above)
Comment (spena) [reply ksqldb-engine/src/main/java/io/confluent/ksql/ddl/commands/DdlCommandExec.java]: True. Done

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/RestoreCommandsCompactorTest.java (hunk 2 above)
Comment (spena) [reply ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/RestoreCommandsCompactorTest.java]: I forgot to delete this file. Done.

File: ksqldb-metastore/src/test/java/io/confluent/ksql/metastore/MetaStoreImplTest.java (hunk 3 above)
Comment (spena) [reply ksqldb-metastore/src/test/java/io/confluent/ksql/metastore/MetaStoreImplTest.java:305]: I added now.

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/RecoveryTest.java (hunk 4 above)
Comment (spena) [reply ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/RecoveryTest.java:862]: Added better comments

**Comment by spena** [comment]
Thanks @jnh5y . I addressed your comments in the last commit.

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/RecoveryTest.java (hunk 4 above)
Comment (jnh5y) [reply ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/RecoveryTest.java:862]: Sorry, to be more precise, I had just noticed that line 862 in RecoveryTest.java says the recovered server only has stream 'B'.

**Review by jnh5y — APPROVED** [review]

**Comment by jnh5y** [comment]
As more nits, checkstyle is unhappy:
```[2022-02-24T18:25:40.240Z] [ERROR] /home/jenkins/workspace/confluentinc-pr_ksql_PR-8803/ksqldb-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java:37:1: Extra separation in import group before 'java.util.Collections' [CustomImportOrder]
[2022-02-24T18:25:40.240Z] [ERROR] /home/jenkins/workspace/confluentinc-pr_ksql_PR-8803/ksqldb-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java:37:1: Wrong lexicographical order for 'java.util.Collections' import. Should be before 'org.slf4j.LoggerFactory'. [CustomImportOrder]
[2022-02-24T18:25:40.240Z] [ERROR] /home/jenkins/workspace/confluentinc-pr_ksql_PR-8803/ksqldb-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java:38:1: Wrong lexicographical order for 'java.util.Iterator' import. Should be before 'org.slf4j.LoggerFactory'. [CustomImportOrder]
[2022-02-24T18:25:40.240Z] [ERROR] /home/jenkins/workspace/confluentinc-pr_ksql_PR-8803/ksqldb-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java:39:1: Wrong lexicographical order for 'java.util.List' import. Should be before 'org.slf4j.LoggerFactory'. [CustomImportOrder]
[2022-02-24T18:25:40.241Z] [ERROR] /home/jenkins/workspace/confluentinc-pr_ksql_PR-8803/ksqldb-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java:40:1: Wrong lexicographical order for 'java.util.Map' import. Should be before 'org.slf4j.LoggerFactory'. [CustomImportOrder]
[2022-02-24T18:25:40.241Z] [ERROR] /home/jenkins/workspace/confluentinc-pr_ksql_PR-8803/ksqldb-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java:41:1: Wrong lexicographical order for 'java.util.Objects' import. Should be before 'org.slf4j.LoggerFactory'. [CustomImportOrder]
[2022-02-24T18:25:40.241Z] [ERROR] /home/jenkins/workspace/confluentinc-pr_ksql_PR-8803/ksqldb-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java:42:1: Wrong lexicographical order for 'java.util.Optional' import. Should be before 'org.slf4j.LoggerFactory'. [CustomImportOrder]
[2022-02-24T18:25:40.241Z] [ERROR] /home/jenkins/workspace/confluentinc-pr_ksql_PR-8803/ksqldb-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java:43:1: Wrong lexicographical order for 'java.util.Set' import. Should be before 'org.slf4j.LoggerFactory'. [CustomImportOrder]
[2022-02-24T18:25:40.242Z] [ERROR] /home/jenkins/workspace/confluentinc-pr_ksql_PR-8803/ksqldb-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java:44:1: Wrong lexicographical order for 'java.util.concurrent.ConcurrentHashMap' import. Should be before 'org.slf4j.LoggerFactory'. [CustomImportOrder]
[2022-02-24T18:25:40.242Z] [ERROR] /home/jenkins/workspace/confluentinc-pr_ksql_PR-8803/ksqldb-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java:45:1: Wrong lexicographical order for 'java.util.stream.Collectors' import. Should be before 'org.slf4j.LoggerFactory'. [CustomImportOrder]
[2022-02-24T18:25:40.242Z] [ERROR] /home/jenkins/workspace/confluentinc-pr_ksql_PR-8803/ksqldb-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java:46:1: Wrong lexicographical order for 'java.util.stream.Stream' import. Should be before 'org.slf4j.LoggerFactory'. [CustomImportOrder]
[2022-02-24T18:25:40.242Z] [ERROR] /home/jenkins/workspace/confluentinc-pr_ksql_PR-8803/ksqldb-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java:47:1: Wrong lexicographical order for 'javax.annotation.concurrent.ThreadSafe' import. Should be before 'org.slf4j.LoggerFactory'. [CustomImportOrder]
[2022-02-24T18:25:40.243Z] [ERROR] /home/jenkins/workspace/confluentinc-pr_ksql_PR-8803/ksqldb-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java:157:95: '+' should be on a new line. [OperatorWrap]
[2022-02-24T18:25:40.243Z] [ERROR] /home/jenkins/workspace/confluentinc-pr_ksql_PR-8803/ksqldb-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java:158:78: '+' should be on a new line. [OperatorWrap]
[2022-02-24T18:25:40.243Z] [ERROR] /home/jenkins/workspace/confluentinc-pr_ksql_PR-8803/ksqldb-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java:159:93: '+' should be on a new line. [OperatorWrap]```

File: ksqldb-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java [hunk 5]
```diff
@@ -117,22 +115,19 @@ public void putSource(final DataSource dataSource,
...
-        info.references.forEach(ref -> {
-          if (ref.equals(dataSource.getName())) {
-            LOG.debug("Add a drop constraint reference back to source '{}' from source '{}'",
-                dataSource.getName().text(), name.text());
-
-            addConstraint(dataSource.getName(), name);
-          }
-        });
+    // Re-build the DROP constraints if existing sources have references to this new source.
+    // This logic makes sure that drop constraints are set back if sources where deleted during
```
[the remark this answers is not in the document]
Comment (spena) [reply ksqldb-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java]: Thanks. I changed the word.

**Pull request merged**